### ISSUE TITLE
Haskell:  Rewrite a term to a constrained symbolic variable

### ISSUE THREAD

**Issue by nishantjr** [kframework/kore#1193]
Whats the correct way to create a symbmolic Int bounded above and below, using the Haskell backend?I'm trying this:

```
module TEST
    imports INT                                                                 
    syntax Interval ::= "[" Int "," Int "]"                                     
                                                                                
    rule [ I , J ] => ?X:Int                                                    
      ensures ?X >=Int I                                                        
      andBool ?X <=Int J                                                        
endmodule 
```

But I see `kore-exec: Quantifying both the term and the predicate probably means that there's
an error somewhere else.`I tried requires as well

**Comment by ttuegel** [comment]
When a rule variable ends up in the term and condition of the configuration, it usually indicates that the backend has tried to execute past the end of the `<k>` cell. We could relax this check so that variables only appearing on the right-hand side of the rule are allowed to appear quantified in the term and condition. @virgil-serbanuta, what do you think?

**Comment by virgil-serbanuta** [comment]
I don't remember all the details about why we didn't want to allow quantifying both the term and the predicate, but what you say seems reasonable regardless of those details.

**Comment by ttuegel** [comment]
So it looks like:

- [ ] [`unwrapAndQuantifyConfiguration`](https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/Step.hs#L162) should fail if it would quantify a variable that only appears on the left-hand side of the configuration 
- [ ] The error condition in [`quantifyPattern`](https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/Simplification/Exists.hs#L352) should be removed.

**Comment by traiansf** [comment]
I started looking into this issue today, and so far, I have two comments (thanks @virgil-serbanuta for helping figuring them out):

1. the current encoding of `?` variables in KORE is incorrect because they are encoded as free variables, which should be implicitly universally quantified, although they are clearly meant to be existential.
2.

**Comment by traiansf** [comment]
Additional comment:

If the above rule is changed into a definitional rule, then it kind of works:

```
module FRESH-VARS
  imports INT

  syntax Int ::= "[" Int "," Int "]" [function]

  rule [ I , J ] => ?X:Int
    ensures ?X >=Int I
    andBool ?X <=Int J

  syntax X ::= "foo"
  rule <k> foo => [3,5] </k>
endmodule
```

Rewriting `foo` produces the output 
```
<k>
  #Exists ?X . ?X
</k>
```

So, no error. The ensures seems to have been lost in the output, but I think this is dues to our imperfect interface with the frontend.

**Comment by ttuegel** [comment]
> The ensures seems to have been lost in the output, but I think this is dues to our imperfect interface with the frontend.

This is definitely a bug.

**Comment by ttuegel** [comment]
See also: #1278

**Comment by nishantjr** [comment]
Any work around I can use for this? @ttuegel

**Comment by nishantjr** [comment]
Would it be possible to use something like `#Exists X. X:Int #and X >=Int I  #and X <=Int J`  directly?

**Comment by ttuegel** [comment]
This will be fixed by #1300.

**Comment by traiansf** [comment]
Just a small comment: #1300 will only fix the backend part, which needs to be propagated and integrated with the Frontend, which should happen in https://github.com/kframework/k/pull/942
